[RDY] Bionics installing overhaul

Night-Pryanik opened the pull request:
Added an alternative for manual and instant installing of CBM. Now you can use autodoc for this task.
- Player has to be on the couch to perform operation;
- ~Autodoc need to be configured prior to use;~
- Autodoc uses player skills and INT to calculate chance of success;
- Installation now takes time, duration is based on bionic difficulty. For example, the easiest CBM (`Power Storage` with difficulty 1) will take twenty minutes to install;
- Player falls asleep because of autodoc's anesthesia (abstracted for now). Player should be sure no monster will disturb the operation;
- Autodoc spawns at the hospital.

Minor: 
- Player will now have 0 sight range when sleeping;
- Added message when player is falling asleep.

[CR] because I want to know if this approach is good enough.

Questions:
- [x] Is autodoc using player skills and INT good? Or should it has some high skill, independent from the player one, removing the need to configure it?
- [x] ~Right now autodoc isn't spawned anywhere. Where should it spawn, and how frequent?~
- [x] Should we add explicit `anesthesia` item, spawn it somewhere, and require to bring it to autodoc? Or we can abstract it like autodoc has it (unlimited?) somewhere in inner storage?
- [x] Is dividing this overhaul to several parts (installing, uninstalling, removing manual (un)installing) good?

John-Candlebury commented:
Ideas for spawning:

-A custom room in labs, similar to the armories and holding cells.
-Custom room in hospitals
-New bionic clinics.
-Black market clinics in house basements.

kevingranade commented:
I can see two major options for installation chance (you essentially have added the first option):
1. Player programs operation into autodoc.
  Outcomes would be based on player int + medical knowledge + programming vs CBM difficulty.
  The time to create this program would scale with CBM install difficulty, and might take a very long time.
   1. Variation, buy a program from a NPC with better skills.
2. Player acquires a program for installing a specific CBM (i.e. on a USB stick).
  This would have an almost 0 chance of bad outcomes.
  This also gives low-int characters a chance of installing, the requirements to just load the software should be very easy.
  Players would find these in lab databases, loaded into autodocs, or perhaps rarely elsewhere on USB sticks.
   1. This might have qualifications for when it will work, such as not working if the targeted body part is mutated.  It will still be safe, but it will refuse to work.

@John-Candlebury covered spawn locations pretty well :+1: 

I'd like to eventually have anesthesia, surgical tools, bandages, suturing material, etc be consumables for this, but we can get it working first and add those things later.

Just getting installation working first is a great approach, it's by far the most common operation.  In particular we don't want to remove the manual operations until this works pretty much perfectly.

inb4 "what about innawoods characters".  Your options will eventually be NPC surgeon or an autodoc (or a mod, sure do whatever in a mod), so innawoods characters with NPCs disabled will be unable to access CBMs, and that's fine.

Night-Pryanik commented:
Added autodoc spawn at the previously empty hospital room, also filled it with lockers and racks.

Night-Pryanik commented:
For now autodoc spawns at the hospital and is working almost fully right (installing bionics still requires painkillers despite autodoc injects anesthesia) for the first step of the overhaul, so I think the PR is ready for the review.

Mecares commented:
Any suggestions how bionics could still be usable in wilderness scenarios after this change? Would be nice to not remove all chances to get bionics from this popular playstyle.

illi-kun commented on src/iexamine.cpp:
```diff
@@ -3497,6 +3498,80 @@ void iexamine::climb_down( player &p, const tripoint &examp )
+
+    uimenu amenu;
+    amenu.selected = 0;
+    amenu.text = string_format( _( "Autodoc Mk. XI. Status: %s. Please choose operation." ), status );
```
```. Please```
Double whitespace required here.

illi-kun commented on src/iexamine.cpp:
```diff
@@ -3497,6 +3498,80 @@ void iexamine::climb_down( player &p, const tripoint &examp )
+    amenu.selected = 0;
+    amenu.text = string_format( _( "Autodoc Mk. XI. Status: %s. Please choose operation." ), status );
+    amenu.addentry( CONFIGURE, true, 'c', str_to_illiterate_str( _( "Configure Autodoc." ) ) );
+    amenu.addentry( INSTALL_CBM, uistate.iexamine_autodoc_config ? true : false, 'i',
```
```uistate.iexamine_autodoc_config ? true : false```
can be effectively shorten to just ```uistate.iexamine_autodoc_config``` since it's already a boolean variable.

illi-kun commented on src/iexamine.cpp:
```diff
@@ -3497,6 +3498,80 @@ void iexamine::climb_down( player &p, const tripoint &examp )
+
+        case INSTALL_CBM: {
+            if( !adjacent_couch ) {
+                popup( _( "No connected couches found. Operation impossible. Exiting." ) );
```
Double whitespaces after periods required.

illi-kun commented on src/iexamine.cpp:
```diff
@@ -3497,6 +3498,80 @@ void iexamine::climb_down( player &p, const tripoint &examp )
+                return;
+            }
+            if( !in_position ) {
+                popup( _( "No patient found located on the connected couches. Operation impossible. Exiting." ) );
```
here as well.

Night-Pryanik commented:
Jenkins rebuild

BevapDin commented on src/iexamine.cpp:
```diff
@@ -3497,6 +3498,80 @@ void iexamine::climb_down( player &p, const tripoint &examp )
+    bool adjacent_couch = false;
+    bool in_position = false;
+    for( const auto &couch_loc : g->m.points_in_radius( examp, 1, 0 ) ) {
+        if( g->m.furn( couch_loc ).obj().id == "f_autodoc_couch" ) {
```
You don't need to get the `furn_t` object to get its id. `map::furn` returns an id that you can compare directly. I suggest you store "f_autodoc_couch" as local constant of type `furn_str_id` and compare it to the result of `map::furn`.

BevapDin commented on src/iexamine.cpp:
```diff
@@ -3497,6 +3498,80 @@ void iexamine::climb_down( player &p, const tripoint &examp )
+            }
+
+            const int bionic = g->inv_for_flag( "CBM", _( "Choose CBM to install" ) );
+            const item &it = p.i_at( bionic );
```
The two references `it` and `itemtype` are only needed below the `if` block, create them there (right before they are needed).

BevapDin commented on src/iexamine.cpp:
```diff
@@ -3497,6 +3498,80 @@ void iexamine::climb_down( player &p, const tripoint &examp )
+
+    switch( static_cast<options>( amenu.ret ) ) {
+        case CONFIGURE:
+            uistate.iexamine_autodoc_config = true;
```
This is a global variable, setting it will affect *all* couches all over the world. It's also not saved anywhere, so it's reset whenever the player loads a game.

Night-Pryanik replied:
> This is a global variable, setting it will affect all couches all over the world.

Can you recommend something to achieve the desired goal, but only for the coach player interacts with?

AMurkin replied:
Prob you need two furnitures: f_autodoc_unconfigured, f_autodoc_configured.

Night-Pryanik replied:
I'd rather not add another furniture. Anybody know, if there is a way to add flags to furniture during runtime?

kevingranade commented:
There is not, that's why there are furniture variants for things like open
door vs closed door.
Your best bet is to drop an item with your state on the same square.

On Tue, Feb 27, 2018, 9:40 PM Anton Burmistrov <notifications@github.com>
wrote:

> *@Night-Pryanik* commented on this pull request.
> ------------------------------
>
> In src/iexamine.cpp
> <https://github.com/CleverRaven/Cataclysm-DDA/pull/22877#discussion_r171147693>
> :
>
> > +    amenu.query();
> +
> +    bool adjacent_couch = false;
> +    bool in_position = false;
> +    for( const auto &couch_loc : g->m.points_in_radius( examp, 1, 0 ) ) {
> +        if( g->m.furn( couch_loc ).obj().id == "f_autodoc_couch" ) {
> +            adjacent_couch = true;
> +            if( p.pos() == couch_loc ) {
> +                in_position = true;
> +            }
> +        }
> +    }
> +
> +    switch( static_cast<options>( amenu.ret ) ) {
> +        case CONFIGURE:
> +            uistate.iexamine_autodoc_config = true;
>
> I'd rather not add another furniture. Anybody know, if there is a way to
> add flags to furniture during runtime?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/CleverRaven/Cataclysm-DDA/pull/22877#discussion_r171147693>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA0gdKSA65l3SNka8kXxVBGMWM9JE2Hsks5tZOa8gaJpZM4R4suj>
> .
>

Night-Pryanik commented:
After player configures the autodoc, the game places pseudo item `autodoc_config` onto the tile with autodoc, which is now behaves like a sealed container with no means of opening. The game then checks for this item, if it exists, then autodoc is considered configured.

BevapDin commented:
> I'd rather not add another furniture. Anybody know, if there is a way to add flags to furniture during runtime?

We don't currently have the infrastructure to store data for a specific place on the map. It's currently solved by having various terrain types (see for example open/closed doors, filled/unfilled kiln).

However, why does the autodoc need to be configure? According to the code, it does nothing, just consumes 3 turns time. This could be done automatically as part of the operation itself. In other words: why are configuration and operation separate steps?

Night-Pryanik commented:
Well, there's no certain reason. The configuration step is for flavor mostly. I just thought that this will help players feel that they're involved in the process. Should I remove the separate configuration step?

kevingranade commented:
Yes remove it. This sounds more complicated than its worth.

Night-Pryanik commented:
Okay, though I have a feeling that we'll have to return it when/if we add a medical software to configure the autodocs.

Night-Pryanik commented:
![default](https://user-images.githubusercontent.com/11132525/37051615-070d78bc-2188-11e8-834f-14a89b3e9bc7.png)

Not too much text at once?

Leland commented:
I think the quantity of text is fine, but I'd probably expect the "Using your knowledge..." text first in that list

Night-Pryanik commented:
That is because the game first checks if player confirms the installation, and then prints text about knowledge and so on.

Night-Pryanik commented:
Moved message about configuring to popup from message in the side log - looks better imo. 
Added message when player falls asleep. 
Made zero sight range when player is asleep.

Night-Pryanik commented:
The `Using your knowledge...` text will appear every time one will install CBM. I think that's not good if this big chunk of text will be repeated time and time again. Maybe bring back the separate configuration step, so the text will be shown only once, and, as I said earlier, it will create base for using the medical software in lieu of manual configuring?

Leland commented:
If the quantity of text is an issue, can be shortened:

`You type into the console, configuring Autodoc to install a CBM`

Night-Pryanik commented:
The "knowledge" part is important, I think it can help players understand that their skills are playing a role in success chance.

Theundyingcode commented:
Was reading through the open PRs out of curiosity but I had a thought on this if it's still getting worked on:

Have the Autodoc "run" the installation program from an SD card just like a CNC machine or 3d printer would. (part of iexamine method) The player could find cards with the "official installation program" from the factory or use a laptop or console to encode cards themselves.  The official factory made onse would result in near zero failure chance while the outcome of player made ones would be subject to the usual int and skill levels.  Requiring a certain level of computers to write the program and/or having it affect the outcomes would be probably be called for in this case.

I'm not familiar with how items are tracked but, given things decay, it should be possible to encode int and relevant skill levels at time of creation.  If not, a rather lengthy copy pasted json file would do the trick.

cainiaowu commented:
This could be changed to [RDY], right? 
After 4 month, I guess everyone already said everything about it and it is already good enough.
I really want to see this get merged.

ZhilkinSerg commented:
Two suggestions for subsequent PRs:

- Make possible to select bionics to install from adjacent tiles instead of from player inventory only;
- Removed need to take painkillers prior to operation as autodoc injects player with anesthetics too.
